### PR TITLE
AUT-1567: Bulk user email audience loader lambda

### PR DESCRIPTION
**Pull request by dbes-gds**
## What?

Bulk user email audience loader lambda.

A new lambda function that when triggered queries the user-profile table to retrieve the required audience for an email campaign, then loads those users in to the bulk-user-email table.

Includes both code and infrastructure, will deploy only to sandpit and build.

This is the first iteration of the loader, subsequent versions will refine the query to return a more specific audience based on a user's terms and conditions version, and potentially the ability to segment loads based on the account creation date so that loads can be split up into several runs rather than one.

This initial version will be used to do some performance testing to see how quickly the loads can take place, then subsequent versions will add improvements and batching if required.

## Why?

We need to create the audience for an email campaign before sending the emails.

## Related PRs

#3227 
#3225

### CODE REVIEW

File: ci/terraform/utils/variables.tf [hunk 1]
```diff
@@ -111,7 +116,17 @@ variable "bulk_user_email_send_schedule_enabled" {
...
+}
+
+variable "bulk_user_email_audience_loader_schedule_enabled" {
+  type    = bool
+  default = false
+}
+
+variable "bulk_user_email_audience_loader_schedule_expression" {
+  type    = string
+  default = "cron(0 13 ? * MON-FRI 2049)"
```
Comment (JHjava) [inline ci/terraform/utils/variables.tf:129]: Could be worth adding a description here and on like 118 to summarize the expression

File: utils/src/main/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public Void handleRequest(ScheduledEvent event, Context context) {
+        LOG.info("Bulk User Email audience load triggered.");
+
+        final long bulkUserEmailMaxAudienceLoadUserCount =
+                configurationService.getBulkUserEmailMaxAudienceLoadUserCount();
+
+        itemCounter = 0;
+        dynamoService
+                .getBulkUserEmailAudienceStream()
```
Comment (JHjava) [inline utils/src/main/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandler.java:56]: I think we only want to add users who didn't accept a certain version of terms and conditions

File: utils/src/test/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import org.junit.jupiter.api.Test;
+import uk.gov.di.authentication.shared.entity.BulkEmailStatus;
+import uk.gov.di.authentication.shared.entity.UserProfile;
+import uk.gov.di.authentication.shared.services.BulkEmailUsersService;
+import uk.gov.di.authentication.shared.services.ConfigurationService;
+import uk.gov.di.authentication.shared.services.DynamoService;
+
+import java.util.List;
+
+import static org.mockito.Mockito.*;
```
Comment (JHjava) [inline utils/src/test/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandlerTest.java:15]: Wildcard import

**Review by JHjava — COMMENTED** [review]
A couple of comments and there are also a couple of code smells

File: utils/src/main/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandler.java (hunk 2 above)
Comment (dbes-gds) [reply utils/src/main/java/uk/gov/di/authentication/utils/lambda/BulkUserEmailAudienceLoaderScheduledEventHandler.java:56]: agreed that's coming in the next version

File: ci/terraform/utils/variables.tf (hunk 1 above)
Comment (dbes-gds) [reply ci/terraform/utils/variables.tf:129]: designed not to trigger and must be overridden, have added a comment, event is not enabled by default

**Review by JHjava — APPROVED** [review]

**Pull request merged**